### PR TITLE
Fix device index issue with XLA + Horovod

### PR DESCRIPTION
**Pull request by kaixih**
The PR is to fix the issue that the XLA always iterates the devices from 0 when creating stream executor on the multi-GPU environment (e.g. Horovod), which might cause none of the devices is picked. For example:

GPU 0: allowed_device=0, device_count=1 => iterate 0 and pick device 0, exceed device_count and exit
GPU 1: allowed_device=1, device_count=1 => iterate 0 (not in allowed_device), exceed device_count and exit (This is wrong!)

We fix this issue by iterating the device indices from the allowed_devices if provided.

GPU 0: allowed_device=0, device_count=1 => iterate 0 and pick device 0, exceed device_count and exit
GPU 1: allowed_device=1, device_count=1 => iterate 1 and pick device 1, exceed device_count and exit

cc. @nluehr

### CODE REVIEW

**Comment by gbaned** [comment]
@r4nt Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
@r4nt Can you please review this PR ? Thank you!

File: tensorflow/compiler/xla/service/platform_util.cc [hunk 1]
```diff
@@ -176,33 +176,41 @@ PlatformUtil::GetStreamExecutors(
...
+    // Once a stream executor is instantiated it will cause allocations on
+    // the device, for example for GPUs cuda context, cudnn handles etc. will
+    // be constructed. By constructing stream executors only on the
+    // allowed_devices, we don't make any allocations on other devices.
+    // This helps in multi-process executions on the same host like horovod or
+    // shared hosts.
+    if (allowed_devices) {
+      int count = 0;
+      for (auto ai = allowed_devices->cbegin(); ai != allowed_devices->cend();
+           ai++) {
```
Comment (pjannaty) [inline tensorflow/compiler/xla/service/platform_util.cc:204]: Would it be possible to use ranged-loop here please? Example https://godbolt.org/z/7sesvzMsM

**Review by pjannaty — COMMENTED** [review]
Thanks, Kaixi! Left a question/comment!

File: tensorflow/compiler/xla/service/platform_util.cc (hunk 1 above)
Comment (kaixih) [reply tensorflow/compiler/xla/service/platform_util.cc]: Thanks. I have pushed new commit to resolve this. PTAL.

**Review by cheshire — APPROVED** [review]

**Pull request merged**